File: python/microsoft-agents-protocol/microsoft/agents/protocol/client/conversation.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""Conversation interface and implementation"""

from abc import ABC, abstractmethod
from typing import Optional, AsyncIterator, TYPE_CHECKING, Dict, Any, List

if TYPE_CHECKING:
    from .simplified_client import SimplifiedClient
    from .stream_event import StreamEvent


class IConversation(ABC):
    """
    Represents a conversation with state maintained across multiple messages.
    """

    @property
    @abstractmethod
    def thread_id(self) -> Optional[str]:
        """Gets the thread ID for this conversation (None until first message sent)"""
        ...

    @property
    @abstractmethod
    def messages(self) -> List[Dict[str, Any]]:
        """
        Gets all messages in this conversation (cached locally).
        Messages are automatically added as the conversation progresses.
        """
        ...

    @abstractmethod
    async def send(self, message: str) -> str:
        """
        Sends a message and returns the complete response as text.

        Args:
            message: The message to send

        Returns:
            The agent's text response
        """
        ...

    @abstractmethod
    async def send_structured(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        Sends a structured message and returns the complete response.

        Args:
            message: The message to send (as dict representation of ChatMessage)

        Returns:
            The agent's response message (as dict)
        """
        ...

    @abstractmethod
    def stream_messages(self, message: str) -> AsyncIterator[Dict[str, Any]]:
        """
        Streams message responses as structured messages (Mode 2: Messages).

        Args:
            message: The message to send

        Returns:
            Async iterator of message dicts
        """
        ...

    @abstractmethod
    def stream_events(self, message: str) -> AsyncIterator["StreamEvent"]:
        """
        Streams raw events (Mode 3: Events).

        Args:
            message: The message to send

        Returns:
            Async iterator of stream events
        """
        ...

    @abstractmethod
    async def get_messages(
        self, limit: Optional[int] = None, after: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Gets all messages from this conversation's thread.

        This is a convenience method that retrieves the full message history
        for this conversation's thread. It delegates to the client's threads API.

        Args:
            limit: Maximum number of messages to return (optional)
            after: Return messages after this message ID (optional)

        Returns:
            List of message dicts in chronological order

        Raises:
            ValueError: If thread_id is not available. Send a message first
                to create a thread.

        Example:
            >>> conversation = client.create_conversation()
            >>> await conversation.send("Hello")
            >>> messages = await conversation.get_messages()
            >>> print(f"Found {len(messages)} messages")

        Note:
            This method requires an active thread. If you haven't sent any
            messages yet, call send() first to create the thread.
        """
        ...


class Conversation(IConversation):
    """
    Internal implementation of IConversation.
    """

    def __init__(
        self,
        client: "SimplifiedClient",
        thread_id: Optional[str],
        enable_logging: bool = False,
        log_directory: str = "logs/conversations"
    ):
        """
        Creates a conversation instance.

        Args:
            client: The client to use for API calls
            thread_id: Optional thread ID to resume
            enable_logging: Enable automatic logging to XML files
            log_directory: Directory path for saving conversation logs
        """
        self._client = client
        self._thread_id = thread_id
        self._messages: List[Dict[str, Any]] = []
        self._enable_logging = enable_logging
        self._log_directory = log_directory

    @property
    def thread_id(self) -> Optional[str]:
        """Gets the thread ID for this conversation"""
        return self._thread_id

    @property
    def messages(self) -> List[Dict[str, Any]]:
        """Gets all messages in this conversation (cached locally)"""
        return self._messages.copy()

    async def send(self, message: str) -> str:
        """
        Sends a message and returns the complete response as text.

        Args:
            message: The message to send

        Returns:
            The agent's text response
        """
        # Create user message
        user_message = {
            "role": "user",
            "contents": [{"kind": "text", "text": message}],
        }

        # Create run request
        request = {
            "thread_id": self._thread_id,
            "input": [user_message],
        }

        # Execute run and wait for completion
        response = await self._client.runs.create_and_wait(request)

        # Update thread ID if this was the first message
        if self._thread_id is None and "thread_id" in response:
            self._thread_id = response["thread_id"]

        # Add user message to cache
        self._messages.append(user_message)

        # Add agent response to cache
        output = response.get("output", [])
        if output:
            for msg in output:
                self._messages.append(msg)

        # Auto-save if logging is enabled
        self._auto_save_conversation()

        # Extract text from agent response
        if not output:
            return ""

        # Find first agent message
        for msg in output:
            if msg.get("role") == "agent":
                contents = msg.get("contents", [])
                for content in contents:
                    if content.get("kind") == "text":
                        return content.get("text", "")

        return ""

    async def send_structured(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        Sends a structured message and returns the complete response.

        Args:
            message: The message to send (as dict representation of ChatMessage)

        Returns:
            The agent's response message (as dict)
        """
        # Create run request
        request = {"thread_id": self._thread_id, "input": [message]}

        # Execute run and wait for completion
        response = await self._client.runs.create_and_wait(request)

        # Update thread ID if this was the first message
        if self._thread_id is None and "thread_id" in response:
            self._thread_id = response["thread_id"]

        # Add user message to cache
        self._messages.append(message)

        # Add agent response to cache
        output = response.get("output", [])
        if output:
            for msg in output:
                self._messages.append(msg)

        # Auto-save if logging is enabled
        self._auto_save_conversation()

        # Extract agent response
        if not output:
            return {"role": "agent", "contents": []}

        # Find first agent message
        for msg in output:
            if msg.get("role") == "agent":
                return msg

        return {"role": "agent", "contents": []}

    async def stream_messages(self, message: str) -> AsyncIterator[Dict[str, Any]]:
        """
        Streams message responses as structured messages (Mode 2: Messages).

        Args:
            message: The message to send

        Yields:
            Message dicts
        """
        # Create run request
        request = {
            "thread_id": self._thread_id,
            "input": [
                {
                    "role": "user",
                    "contents": [{"kind": "text", "text": message}],
                }
            ],
        }

        # Track messages by ID
        message_map: Dict[str, Dict[str, Any]] = {}

        # Stream events
        async for evt in self._client._stream_run(request):
            # Update thread ID from first event
            if self._thread_id is None and evt["event_type"] == "run.started":
                run_data = evt.get("data", {})
                if "thread_id" in run_data:
                    self._thread_id = run_data["thread_id"]

            # Handle message events
            if evt["event_type"] == "message.created":
                message_data = evt.get("data", {})
                message_id = message_data.get("message_id")
                if message_id:
                    message_map[message_id] = message_data
                    yield message_data
            elif evt["event_type"] in ("message.updated", "message.delta"):
                message_data = evt.get("data", {})
                message_id = message_data.get("message_id")
                if message_id:
                    message_map[message_id] = message_data
                    yield message_data

    async def stream_events(self, message: str) -> AsyncIterator["StreamEvent"]:
        """
        Streams raw events (Mode 3: Events).

        Args:
            message: The message to send

        Yields:
            Stream events
        """
        from .stream_event import StreamEvent

        # Create run request
        request = {
            "thread_id": self._thread_id,
            "input": [
                {
                    "role": "user",
                    "contents": [{"kind": "text", "text": message}],
                }
            ],
        }

        # Stream events
        async for evt in self._client._stream_run(request):
            # Update thread ID from first event
            if self._thread_id is None and evt["event_type"] == "run.started":
                run_data = evt.get("data", {})
                if "thread_id" in run_data:
                    self._thread_id = run_data["thread_id"]

            yield StreamEvent(event_type=evt["event_type"], data=evt.get("data", {}))

    async def get_messages(
        self, limit: Optional[int] = None, after: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Gets all messages from this conversation's thread.

        This is a convenience method that retrieves the full message history
        for this conversation's thread. It delegates to the client's threads API.

        Args:
            limit: Maximum number of messages to return (optional)
            after: Return messages after this message ID (optional)

        Returns:
            List of message dicts in chronological order

        Raises:
            ValueError: If thread_id is not available. Send a message first
                to create a thread.

        Example:
            >>> conversation = client.create_conversation()
            >>> await conversation.send("Hello")
            >>> messages = await conversation.get_messages()
            >>> print(f"Found {len(messages)} messages")

        Note:
            This method requires an active thread. If you haven't sent any
            messages yet, call send() first to create the thread.
        """
        if not self.thread_id:
            raise ValueError(
                "No thread ID available. Send a message to this conversation "
                "first to create a thread."
            )

        return await self._client.threads.get_messages(
            thread_id=self.thread_id, limit=limit, after=after
        )

    def _auto_save_conversation(self) -> None:
        """Automatically saves the conversation to XML if logging is enabled."""
        if not self._enable_logging or not self._thread_id:
            return

        try:
            from pathlib import Path

            # Ensure log directory exists
            log_path = Path(self._log_directory)
            log_path.mkdir(parents=True, exist_ok=True)

            # Save conversation to file
            file_path = log_path / f"{self._thread_id}.xml"
            file_path.write_text(str(self))
        except Exception:
            # Silently ignore logging errors to avoid breaking the main flow
            pass

    def __str__(self) -> str:
        """
        Returns the XML representation of all messages in this conversation.

        Returns:
            XML string with all messages wrapped in a thread element
        """
        if not self._messages:
            return '<?xml version="1.0" encoding="utf-8"?>\n<thread />'

        # Build simple XML representation
        lines = ['<?xml version="1.0" encoding="utf-8"?>', '<thread>']

        for msg in self._messages:
            role = msg.get("role", "unknown")
            lines.append(f'  <{role}>')

            contents = msg.get("contents", [])
            for content in contents:
                kind = content.get("kind", "")
                if kind == "text":
                    text = content.get("text", "")
                    # Escape XML special characters
                    text = text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
                    lines.append(f'    <text>{text}</text>')

            lines.append(f'  </{role}>')

        lines.append('</thread>')
        return '\n'.join(lines)

File: typescript/packages/agents-protocol-client/src/conversation.ts
/**
 * Conversation interface and implementation for stateful multi-turn chats
 */

import type { ChatMessage, ChatRole } from '@microsoft/agents-protocol-abstractions';
import type { SimplifiedClient } from './simplified-client';
import type { StreamEvent } from './stream-event';
import { MessageSerializer } from '@microsoft/agents-xml';

/**
 * Represents a conversation with state maintained across multiple messages
 */
export interface IConversation {
  /**
   * Gets the thread ID for this conversation (null until first message sent)
   */
  readonly threadId?: string;

  /**
   * Gets all messages in this conversation (cached locally).
   * Messages are automatically added as the conversation progresses.
   */
  readonly messages: ChatMessage[];

  /**
   * Sends a message and returns the complete response as text
   * @param message The message to send
   * @param signal Optional abort signal for cancellation
   * @returns The agent's text response
   */
  send(message: string, signal?: AbortSignal): Promise<string>;

  /**
   * Sends a structured message and returns the complete response
   * @param message The message to send
   * @param signal Optional abort signal for cancellation
   * @returns The agent's response message
   */
  sendStructured(message: ChatMessage, signal?: AbortSignal): Promise<ChatMessage>;

  /**
   * Streams message responses as structured messages
   * @param message The message to send
   * @param signal Optional abort signal for cancellation
   * @returns Async generator of messages
   */
  streamMessages(
    message: string,
    signal?: AbortSignal
  ): AsyncGenerator<ChatMessage, void, undefined>;

  /**
   * Streams raw events from the conversation
   * @param message The message to send
   * @param signal Optional abort signal for cancellation
   * @returns Async generator of stream events
   */
  streamEvents(
    message: string,
    signal?: AbortSignal
  ): AsyncGenerator<StreamEvent, void, undefined>;

  /**
   * Gets all messages from this conversation's thread.
   *
   * This is a convenience method that retrieves the full message history
   * for this conversation's thread. It delegates to the client's threads API.
   *
   * @param options - Optional parameters for message retrieval
   * @param options.limit - Maximum number of messages to return
   * @param options.after - Return messages after this message ID
   * @returns Promise resolving to array of ChatMessage objects in chronological order
   * @throws {Error} If thread ID is not available. Send a message first to create a thread.
   *
   * @example
   * ```typescript
   * const conversation = client.createConversation();
   * await conversation.send("Hello");
   * const messages = await conversation.getMessages();
   * console.log(`Found ${messages.length} messages`);
   * ```
   *
   * @remarks
   * This method requires an active thread. If you haven't sent any messages yet,
   * call send() first to create the thread.
   */
  getMessages(options?: {
    limit?: number;
    after?: string;
  }): Promise<ChatMessage[]>;
}

/**
 * Internal implementation of IConversation
 */
export class Conversation implements IConversation {
  private _threadId?: string;
  private _messages: ChatMessage[] = [];
  private readonly _enableLogging: boolean;
  private readonly _logDirectory: string;

  constructor(
    private readonly client: SimplifiedClient,
    threadId?: string,
    enableLogging: boolean = false,
    logDirectory: string = 'logs/conversations'
  ) {
    this._threadId = threadId;
    this._enableLogging = enableLogging;
    this._logDirectory = logDirectory;
  }

  get threadId(): string | undefined {
    return this._threadId;
  }

  get messages(): ChatMessage[] {
    return [...this._messages];
  }

  /**
   * Sends a text message and returns the agent's text response
   */
  async send(message: string, signal?: AbortSignal): Promise<string> {
    const userMessage: ChatMessage = {
      role: 'user' as ChatRole,
      messageId: this.generateMessageId(),
      contents: [
        {
          kind: 'text',
          type: 'text',
          text: message,
        } as any,
      ],
    };

    const request = {
      threadId: this._threadId,
      input: [userMessage],
    };

    // Use internal client method to create run
    const response = await this.client.createRunAndWait(request, signal);

    // Update thread ID if this was the first message
    if (!this._threadId && response.threadId) {
      this._threadId = response.threadId;
    }

    // Add user message to cache
    this._messages.push(userMessage);

    // Add agent response to cache
    if (response.output && response.output.length > 0) {
      for (const msg of response.output) {
        this._messages.push(msg);
      }
    }

    // Auto-save if logging is enabled
    this.autoSaveConversation();

    // Extract text from agent response
    if (!response.output || response.output.length === 0) {
      return '';
    }

    const agentMessage = response.output.find((m) => m.role === 'agent');
    if (!agentMessage) {
      return '';
    }

    return this.extractText(agentMessage);
  }

  /**
   * Sends a structured message and returns the agent's structured response
   */
  async sendStructured(
    message: ChatMessage,
    signal?: AbortSignal
  ): Promise<ChatMessage> {
    const request = {
      threadId: this._threadId,
      input: [message],
    };

    const response = await this.client.createRunAndWait(request, signal);

    // Update thread ID if this was the first message
    if (!this._threadId && response.threadId) {
      this._threadId = response.threadId;
    }

    // Add user message to cache
    this._messages.push(message);

    // Add agent response to cache
    if (response.output && response.output.length > 0) {
      for (const msg of response.output) {
        this._messages.push(msg);
      }
    }

    // Auto-save if logging is enabled
    this.autoSaveConversation();

    // Return first agent message
    if (!response.output || response.output.length === 0) {
      return {
        role: 'agent' as ChatRole,
        messageId: this.generateMessageId(),
        contents: [],
      };
    }

    return (
      response.output.find((m) => m.role === 'agent') ?? {
        role: 'agent' as ChatRole,
        messageId: this.generateMessageId(),
        contents: [],
      }
    );
  }

  /**
   * Streams messages from the agent
   */
  async *streamMessages(
    message: string,
    signal?: AbortSignal
  ): AsyncGenerator<ChatMessage, void, undefined> {
    const userMessage: ChatMessage = {
      role: 'user' as ChatRole,
      messageId: this.generateMessageId(),
      contents: [
        {
          kind: 'text',
          type: 'text',
          text: message,
        } as any,
      ],
    };

    const request = {
      threadId: this._threadId,
      input: [userMessage],
    };

    // Track messages by ID
    const messageMap = new Map<string, ChatMessage>();

    // Stream events from client
    for await (const event of this.client.streamRun(request, signal)) {
      // Update thread ID from first event
      if (!this._threadId && event.eventType === 'run.started') {
        const runData = event.data as any;
        if (runData?.threadId) {
          this._threadId = runData.threadId;
        }
      }

      // Handle message events
      if (event.eventType === 'message.created') {
        const messageData = event.data as ChatMessage;
        if (messageData?.messageId) {
          messageMap.set(messageData.messageId, messageData);
          yield messageData;
        }
      } else if (
        event.eventType === 'message.updated' ||
        event.eventType === 'message.delta'
      ) {
        const messageData = event.data as ChatMessage;
        if (messageData?.messageId) {
          messageMap.set(messageData.messageId, messageData);
          yield messageData;
        }
      }
    }
  }

  /**
   * Streams raw events from the agent
   */
  async *streamEvents(
    message: string,
    signal?: AbortSignal
  ): AsyncGenerator<StreamEvent, void, undefined> {
    const userMessage: ChatMessage = {
      role: 'user' as ChatRole,
      messageId: this.generateMessageId(),
      contents: [
        {
          kind: 'text',
          type: 'text',
          text: message,
        } as any,
      ],
    };

    const request = {
      threadId: this._threadId,
      input: [userMessage],
    };

    // Stream all events from client
    for await (const event of this.client.streamRun(request, signal)) {
      // Update thread ID from first event
      if (!this._threadId && event.eventType === 'run.started') {
        const runData = event.data as any;
        if (runData?.threadId) {
          this._threadId = runData.threadId;
        }
      }

      yield event;
    }
  }

  /**
   * Extracts text content from a message
   */
  private extractText(message: ChatMessage): string {
    if (!message.contents || message.contents.length === 0) {
      return '';
    }

    const textContent = message.contents.find(
      (c: any) => c.kind === 'text'
    ) as any;

    return textContent?.text ?? '';
  }

  /**
   * Gets all messages from this conversation's thread
   */
  async getMessages(options?: {
    limit?: number;
    after?: string;
  }): Promise<ChatMessage[]> {
    if (!this._threadId) {
      throw new Error(
        'No thread ID available. Send a message to this conversation first to create a thread.'
      );
    }

    return await this.client.threads.getMessages(this._threadId, options);
  }

  /**
   * Generates a unique message ID
   */
  private generateMessageId(): string {
    return `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Automatically saves the conversation to XML if logging is enabled
   */
  private autoSaveConversation(): void {
    if (!this._enableLogging || !this._threadId) {
      return;
    }

    try {
      const fs = require('fs');
      const path = require('path');

      // Ensure log directory exists
      if (!fs.existsSync(this._logDirectory)) {
        fs.mkdirSync(this._logDirectory, { recursive: true });
      }

      // Save conversation to file
      const filePath = path.join(this._logDirectory, `${this._threadId}.xml`);
      fs.writeFileSync(filePath, this.toString(), 'utf-8');
    } catch {
      // Silently ignore logging errors to avoid breaking the main flow
    }
  }

  /**
   * Returns the XML representation of all messages in this conversation
   * @returns XML string with all messages wrapped in a thread element
   */
  toString(): string {
    if (this._messages.length === 0) {
      return '<?xml version="1.0" encoding="utf-8"?>\n<thread />';
    }

    const serializer = new MessageSerializer();
    const messagesXml = serializer.serializeMany(this._messages);

    return `<?xml version="1.0" encoding="utf-8"?>\n<thread>\n${messagesXml}\n</thread>`;
  }
}
